source/agora/node/EnrollmentManager.d

```diff
@@ -0,0 +1,138 @@
+/*******************************************************************************
+    Contains supporting code for tracking the current ledger.
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+    License:
+        MIT License. See LICENSE for details.
+*******************************************************************************/
+
+module agora.node.EnrollmentManager;
+
+import agora.common.Amount;
+import agora.common.crypto.ECC;
+import agora.common.crypto.Key;
+import agora.common.Hash;
+import agora.common.Types;
+import agora.common.crypto.Schnorr : Pair;
+
+import std.random;
+import std.stdio;
+
+/*******************************************************************************
+    Defines data for enrollment process
+*******************************************************************************/
+
+public struct EnrollmentData{
+    /// K: UTXO key, A public key matching a frozen UTXO
+    public PublicKey K;
+
+    /// X: random seed, The nth image of random value
+    public Hash X;
+
+    /// n: cycle length, A number within bounds [0; 1008]
+    /// original spec: [0; (Balance(K) % 40,000) * freezing period * 2]
+    public ulong n;
+
+    /// R: signature noise, The initial nonce used for signing
+    public Point R;
+
+    /// S: A signature for the messge H(K, X, n, R) and the key K, using R
+    public Signature S;
+
+    /***************************************************************************
+        Implements hashing support
+        Params:
+            dg = hashing function
+    ***************************************************************************/
+    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
+    {
+        hashPart(this.K, dg);
+        hashPart(this.X, dg);
+        hashPart(this.n, dg);
+        hashPart(this.R, dg);
+    }
+}
+
+/*******************************************************************************
+    Handles data and process for enrollment
+*******************************************************************************/
+
+public struct EnrollmentManager
+{
+    /// Node's key pair
+    private KeyPair node_keypair;
+
+    /// Random key for enrollment
+    private Pair random_key;
+
+    /// Node balance
+    private Amount balance;
+
+    // Random object
+    private Random genint;
+
+    /// EnrollmentData
+    private EnrollmentData* data;
+
+    /***************************************************************************
+        Constructor
+        Params:
+            keypair = the keypair of the owner node
+            balance = account balance of the owner node
+    ***************************************************************************/
+    public this(KeyPair keypair, Amount balance)
+    {
+        this.node_keypair = keypair;
+        this.balance = balance;
+        this.genint = Random(123);
```

Comment:
This is not a safe way to use a random-number generator. The `123` part is the seed. The sequence of numbers you get out of a random-number generator is always the same for the given seed.

Instead, I suggest using `randombytes_uniform()` from libsodium to generate a number within a given range.